docs/proposed/http-storage-node-protocol.rst

```diff
@@ -13,6 +13,54 @@ Specifically, it should be possible to implement a Tahoe-LAFS storage server wit
 The Tahoe-LAFS client will also need to change but it is not expected that it will be noticably simplified by this change
 (though this may be the first step towards simplifying it).
 
+Motivation
+----------
+
+Foolscap
+~~~~~~~~
+
+Foolscap is a remote method invocation protocol with several distinctive features.
+At its core it allows separate processes to refer each other's objects and methods using a capability-based model.
+This allows for extremely fine-grained access control in a system that remains highly securable without becoming overwhelmingly complicated.
+Supporting this is a flexible and extensible serialization system which allows data to be exchanged between processes in carefully controlled ways.
+
+Tahoe avails itself of only a small portion of these features.
+A Tahoe storage server typically only exposes one object with a fixed set of methods to clients.
+A Tahoe introducer node does roughly the same.
+Tahoe exchanges simple data structures that have many common, standard serialized representations.
+
+In exchange for this slight use of Foolscap's sophisticated mechanisms,
+Tahoe pays a substantial price:
+
+* Foolscap is implemented only for Python.
+  Tahoe is thus limited to being implemented on in Python.
+* There is only one Python implementation of Foolscap.
+  The implementation is therefore the de facto standard and understanding of the protocol often relies on understanding that implementation.
+* The Foolscap developer community is very small.
+  The implementation therefore advances very little and some non-trivial part of the maintenance cost falls on the Tahoe project.
+* The extensible serialization system imposes substantial overhead for the simple data structures Tahoe exchanges.
+  Tahoe therefore presents a more sluggish experience to users and taxes servers more greatly than is necessary.
```

Comment:
Is "speed" really the concern here? (Like, is it really slower than JSON or something? Also does that matter?) Seems like real concern is more "it's big and flexible and thus complex for developers, but Tahoe doesn't need that"

Reply:
Added some more justification for this conclusion (but without concrete benchmarks I'm not going to claim *too* much confidence/authority here... If you're still skeptical we can drop this part).

Reply:
> Perhaps just saying "We believe an HTTP-based protocol ... faster"?

That seems like a reasonable hedge, given the lack of evidence. :)

> Also, do we even care? If our HTTP implementation turned out to be slower, would we go back to Foolscap? (I think: "no").

I think we *care* even if making things faster isn't the primary objective.

On the server-side, I think the performance benefits are real because a 5% reduction in CPU overhead (say, from serialization) translates directly in to a 5% increase in transfer capacity from the same resources.

Reply:
I mean "care" in the context of GBS: I'm not saying performance-benefits are bad .. I'm saying that I think we should prefer a simpler, easier-to-understand implementation *even if* if doesn't have performance benefits. That is, while "it's faster" would certainly be a cool side-effect I'd take slower too (if that meant easier to understand/implement/interop-with).

Maybe even this little thread shows that we can get side-tracked about performance in this regard ;)